services/grpc/handler.go

```diff
@@ -48,21 +49,63 @@ func (h *Handler) SendEvent(ctx context.Context, req *pb.SendEventRequest) (*pb.
 	metrics.Increment("batches_read_total", fmt.Sprintf("status=success,conn_group=%s", identifier.Group))
 	h.sendEventCounters(req.Events, identifier.Group)
 
+	responseChannel := make(chan *pb.SendEventResponse, 1)
 	h.C.Collect(ctx, &collection.CollectRequest{
 		ConnectionIdentifier: identifier,
 		TimeConsumed:         timeConsumed,
 		SendEventRequest:     req,
+		AckFunc:              h.Ack(responseChannel, req.ReqGuid, identifier.Group),
 	})
+	return <-responseChannel, nil
 
-	return &pb.SendEventResponse{
-		Status:   pb.Status_STATUS_SUCCESS,
-		Code:     pb.Code_CODE_OK,
-		SentTime: time.Now().Unix(),
-		Data: map[string]string{
-			"req_guid": req.GetReqGuid(),
-		},
-	}, nil
+}
 
+func (h *Handler) Ack(responseChannel chan *pb.SendEventResponse, reqGuid, connGroup string) collection.AckFunc {
+	switch config.Event.Ack {
+	case 0:
+		responseChannel <- &pb.SendEventResponse{
+			Status:   pb.Status_STATUS_SUCCESS,
+			Code:     pb.Code_CODE_OK,
+			SentTime: time.Now().Unix(),
+			Data: map[string]string{
+				"req_guid": reqGuid,
+			},
+		}
+		return nil
+	case 1:
+		return func(err error) {
+			if err != nil {
+				logger.Error(fmt.Sprintf("[grpc.Ack] publish message failed for %s: %v", connGroup, err))
```

Comment:
(suggestions): Instead of specifying formatting everywhere with `fmt.Sprintf` we could use `logger.Errorf`

Reply:
Updated, let's update this throughout Raccoon, part from current PR.